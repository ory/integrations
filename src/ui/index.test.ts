import { UiNode } from '@ory/client'
import { filterNodesByGroups, getNodeLabel } from './index'
import nodes from './fixtures/nodes.json'

describe('generic helpers', () => {
  const testNodes: Array<any> = [
    { group: 'default', type: 'a' },
    { group: 'foo', type: 'b' },
    { group: 'bar', type: 'c' },
    {
      group: 'default',
      attributes: {
        name: 'a',
        type: 'checkbox',
        node_type: 'input'
      }
    },
    {
      group: 'foo',
      attributes: {
        name: 'c',
        type: 'hidden',
        node_type: 'input'
      }
    },
    {
      group: 'webauthn',
      attributes: {
        name: 'd',
        type: 'script',
        node_type: 'input'
      }
    },
    {
      group: 'webauthn',
      attributes: {
        name: 'e',
        type: 'input',
        node_type: 'input'
      }
    },
    {
      group: 'totp',
      attributes: {
        name: 'f',
        type: 'input',
        node_type: 'input'
      }
    }
  ]

  const uiNodes = testNodes as Array<UiNode>

  const tc: Array<{
    description: string
    opts: Partial<filterNodesByGroups>
    expected: Array<any>
  }> = [
    {
      description: 'nodes with the checkbox attribute',
      opts: {
        attributes: 'checkbox',
        withoutDefaultAttributes: true
      },
      expected: [
        {
          group: 'default',
          attributes: {
            name: 'a',
            type: 'checkbox',
            node_type: 'input'
          }
        }
      ]
    },
    {
      description:
        'filtering by nodes should always include the default attributes',
      opts: {
        attributes: 'email',
        withoutDefaultAttributes: false
      },
      expected: [
        {
          group: 'webauthn',
          attributes: {
            name: 'd',
            type: 'script',
            node_type: 'input'
          }
        },
        {
          group: 'webauthn',
          attributes: {
            name: 'e',
            type: 'input',
            node_type: 'input'
          }
        },
        {
          group: 'totp',
          attributes: {
            name: 'f',
            type: 'input',
            node_type: 'input'
          }
        }
      ]
    },
    {
      description:
        'can filter by nodes using comma seperated groups and should always include the default groups',
      opts: {
        groups: 'foo,bar'
      },
      expected: [
        { group: 'default', type: 'a' },
        { group: 'foo', type: 'b' },
        { group: 'bar', type: 'c' },
        {
          group: 'default',
          attributes: {
            name: 'a',
            type: 'checkbox',
            node_type: 'input'
          }
        },
        {
          group: 'foo',
          attributes: {
            name: 'c',
            type: 'hidden',
            node_type: 'input'
          }
        }
      ]
    },
    {
      description:
        'can filter by nodes using an array of groups and should always include the default groups',
      opts: {
        groups: ['foo', 'bar']
      },
      expected: [
        { group: 'default', type: 'a' },
        { group: 'foo', type: 'b' },
        { group: 'bar', type: 'c' },
        {
          group: 'default',
          attributes: {
            name: 'a',
            type: 'checkbox',
            node_type: 'input'
          }
        },
        {
          group: 'foo',
          attributes: {
            name: 'c',
            type: 'hidden',
            node_type: 'input'
          }
        }
      ]
    },
    {
      description:
        'can filter by nodes using a group and exclude default groups',
      opts: {
        groups: ['foo'],
        withoutDefaultGroup: true
      },
      expected: [
        { group: 'foo', type: 'b' },
        {
          group: 'foo',
          attributes: {
            name: 'c',
            type: 'hidden',
            node_type: 'input'
          }
        }
      ]
    },
    {
      description: 'can filter by attributes and exclude default attributes',
      opts: {
        attributes: 'hidden',
        withoutDefaultAttributes: true
      },
      expected: [
        {
          group: 'foo',
          attributes: {
            name: 'c',
            type: 'hidden',
            node_type: 'input'
          }
        }
      ]
    },
    {
      description: 'can filter by attributes and exclude another attributes',
      opts: {
        groups: 'webauthn',
        withoutDefaultGroup: true,
        excludeAttributes: 'script'
      },
      expected: [
        {
          group: 'webauthn',
          attributes: {
            name: 'e',
            type: 'input',
            node_type: 'input'
          }
        }
      ]
    }
  ]

  tc.forEach(({ description, opts, expected }) => {
    test(description, () => {
      expect(
        filterNodesByGroups({
          nodes: uiNodes,
          ...opts
        })
      ).toEqual(expected)
    })
  })

  test('getNodeLabel', () => {
    expect(nodes.map(getNodeLabel)).toMatchSnapshot()
  })
})
