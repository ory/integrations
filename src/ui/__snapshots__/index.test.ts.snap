// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generic helpers getNodeLabel 1`] = `
[
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "Save",
  "Password",
  "Save",
  "",
  "Remove security key "bar"",
  "Remove security key "foo"",
  "Name of the security key",
  "Add security key",
  "",
  "",
  "Authenticator app QR code",
  "This is your authenticator app secret. Use it if you can not scan the QR code.",
  "Verify code",
  "Save",
]
`;
